### DISCUSSION TITLE
How to make a fort.15 file for NWS = 30 Blended GAHM & Background Gridded Wind and Pressure?

### DISCUSSION THREAD

**Discussion by jalal-90** [StormSurgeLive/asgs#1205] [General]
Dear sir,

I use the following command in perl to produce fort.15 file for nws 30.
perl control_file_gen.pl [--cst 2012092100] [--dt 10] [--nws 30] [--nowcast] [--controltemplate fort_template.15] [--gridname fort.14] < fort.22

But I am getting the following errors.
Use of uninitialized value $advisdir in concatenation (.) or string at control_file_gen.pl line 208.
Use of uninitialized value $enstorm in concatenation (.) or string at control_file_gen.pl line 208.
Use of uninitialized value $csdate in pattern match (m//) at control_file_gen.pl line 228.
Use of uninitialized value $enstorm in concatenation (.) or string at control_file_gen.pl line 1371.
[2023-Sep-26-T18:02:44] DEBUG: : control_file_gen.pl: nws is 0 and waves digit is 0.
Use of uninitialized value $controltemplate in concatenation (.) or string at control_file_gen.pl line 258.
Use of uninitialized value $controltemplate in concatenation (.) or string at control_file_gen.pl line 259.
Use of uninitialized value $enstorm in concatenation (.) or string at control_file_gen.pl line 1371.
[2023-Sep-26-T18:02:44] ERROR: : control_file_gen.pl: Failed to open the fort.15 template file  for reading: No such file or directory.
Died at control_file_gen.pl line 260.

Any suggestions would be appreciated. I added the code and input files herewith.

[files.zip](https://github.com/StormSurgeLive/asgs/files/12725884/files.zip)

**Comment by wwlwpd** [comment]
Hi @jalal-90 - I converted this to a discussion; trying getting rid of the square braces (`[` and `]`) and let us know how it works.

**Reply by jalal-90** [reply]
I removed square braces from the command line and found the following errors.

perl control_file_gen.pl --cst 2012092100 --dt 10 --nws 30 --controltemplate fort_template.15 --gridname fort.14 < fort.22

errors:
Use of uninitialized value $advisdir in concatenation (.) or string at control_file_gen.pl line 208.
Use of uninitialized value $enstorm in concatenation (.) or string at control_file_gen.pl line 208.
Use of uninitialized value $enstorm in concatenation (.) or string at control_file_gen.pl line 1371.
[2023-Sep-27-T09:30:50] DEBUG: : control_file_gen.pl: nws is 30 and waves digit is 0.
Use of uninitialized value $enstorm in concatenation (.) or string at control_file_gen.pl line 1371.
[2023-Sep-27-T09:30:50] ERROR: : control_file_gen.pl: Failed to open the output control file //fort.15: Permission denied
Died at control_file_gen.pl line 266.

**Comment by jasonfleming** [comment]
The `control_file_gen.pl` script is trying to place your `fort.15` file in the proper directory, which it expected to be named according to the storm scenario and the advisory number. For example, if your scenario is named 'myNWS30track' and the advisory number is '10', then it wants to create a file called '10/myNWS30track/fort.15'. You can solve this by providing an advisory number with the `--advisdir` command line option and a scenario name with the `--enstorm` option. 

I also see that it could not find the fort.15 template file that you specified on the command line. Do you have a fort.15 template file (or a working `fort.15` file) for you mesh?

**Comment by jalal-90** [comment]
Thank you so much, sir, for your kind advice. No, I do not have a fort.15 template file. Could I have it?

Another humble request, could I have a tool or script that can produce a fort.221 and a fort.222 file for nws 30?

**Comment by jasonfleming** [comment]
Yes, we have many `fort.15` template files, but they are specific to each mesh (`fort.14`) and nodal attributes file (`fort.13`) that we use.  Do you already have a mesh and nodal attributes file that you'd like to use?

The NWS30 (blended OWI win/pre and GAHM) feature of ADCIRC requires that you have an ATCF-formatted track file as well as an OWI win/pre  dataset (`fort.22`/`fort.221`/`fort.222`). We have scripts that generate those files, but they currently assume that live data sources are being used. Do you already have the track file and gridded meteorological data that you'd like to use?

**Reply by jalal-90** [reply]
Yes, sir. I have a mesh and nodal attributes files. I have also the track file and gridded meteorological data.

**Reply by jalal-90** [reply]
Sir, you said that you have scripts that generate fort.221 and fort.222 files. Could I have those scripts?

**Reply by jasonfleming** [reply]
Yes, they are in this repository, depending on the source of the data (NAM or GFS). But it sounds like you already have the gridded wind fields (i.e., fort.221, fort.222). Can you be more specific?

**Reply by jalal-90** [reply]
Sir, I have gridded wind (u and v) and mean sea level pressure from ERA5 reanalysis data. I have written a MATLAB code to generate fort.221 and fort.222 files. But I am getting the following error after running the ADCIRC (version 55.01) model.

ERROR: check_owi_err: Failed to 'reading wind u-velocity snapshot, domain 1' from './fort.222' (unit number 222).
 iostat =          64
ERROR: check_owi_err: ADCIRC execution terminated.

![1696210368938](https://github.com/StormSurgeLive/asgs/assets/139944952/00f75966-396f-400d-b8ad-d2c92df8e95b)

**Comment by jasonfleming** [comment]
Do you have a `fort.15` file that you have already used with your mesh and nodal attributes file?

**Reply by jalal-90** [reply]
Yes, sir. I have a fort.15 file. 

I can run the ADCIRC model for NWS =20. Now, I have a problem with formatting fort.221, fort.222, and fort.15 files for NWS =30.

**Comment by jasonfleming** [comment]
What documentation did you use to learn the format of the `fort.221` and `fort.222`? Can you show me what the first 5 lines of your `fort.221` and `fort.222` look like?

**Reply by jalal-90** [reply]
From the following link, I learned the format of the fort.221 and fort.222 files.
https://wiki.adcirc.org/NWS12

Here is the first 5 lines of fort.221 and fort.222 files.
![1696267841078](https://github.com/StormSurgeLive/asgs/assets/139944952/707634e6-3eaa-463d-b478-813d55e4d289)

![1696267907996](https://github.com/StormSurgeLive/asgs/assets/139944952/f3bf983b-a86e-4308-b940-d6e9b1072c53)

**Reply by jalal-90** [reply]
I also follow the format from here.
https://github.com/adcirc/adcirc-testsuite/tree/master/adcirc/adcirc_nws30_wlcorrection

**Comment by wwlwpd** [comment]
@jalal-90 this might go faster if you join our dicord chatroom we created specifically to sort out this issue, https://discord.gg/YWH5xPNt (link expires in 7 days). Thank you!

**Reply by jalal-90** [reply]
I joined there. Thank you so much.
